components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1233 @@
+        let rows = stmt.query_and_then(NO_PARAMS, |row| row.get_checked::<_, SyncGuid>("guid"))?;
+        for row in rows {
+            let guid = row?;
+            tree.note_deleted(guid.into());
+        }
+
+        Ok(tree)
+    }
+
+    /// Fetches content info for all synced items that changed since the last
+    /// sync and don't exist locally.
+    fn fetch_new_remote_contents(&self) -> Result<HashMap<dogear::Guid, Content>> {
+        let mut contents = HashMap::new();
+
+        let sql = format!(
+            r#"
+            SELECT v.guid, v.kind, IFNULL(v.title, "") AS title, h.url,
+                   s.position
+            FROM moz_bookmarks_synced v
+            JOIN moz_bookmarks_synced_structure s ON s.guid = v.guid
+            LEFT JOIN moz_places h ON h.id = v.placeId
+            LEFT JOIN moz_bookmarks b ON b.guid = v.guid
+            WHERE NOT v.isDeleted AND
+                  v.needsMerge AND
+                  b.guid IS NULL AND
+                  s.parentGuid <> '{root_guid}'"#,
+            root_guid = BookmarkRootGuid::Root.as_guid().as_ref()
+        );
+        let mut stmt = self.db.prepare(&sql)?;
+        let mut results = stmt.query(NO_PARAMS)?;
+        while let Some(result) = results.next() {
+            let row = result?;
+            let content = match SyncedBookmarkKind::from_u8(row.get_checked("kind")?)? {
+                SyncedBookmarkKind::Bookmark | SyncedBookmarkKind::Query => {
+                    let title = row.get_checked("title")?;
+                    let url_href = row.get_checked("url")?;
+                    Content::Bookmark { title, url_href }
+                }
+                SyncedBookmarkKind::Folder => {
+                    let title = row.get_checked("title")?;
+                    Content::Folder { title }
+                }
+                SyncedBookmarkKind::Separator => {
+                    let position = row.get_checked("position")?;
+                    Content::Separator { position }
+                }
+                _ => continue,
+            };
+            let guid = row.get_checked::<_, SyncGuid>("guid")?;
+            contents.insert(guid.into(), content);
+        }
+
+        Ok(contents)
+    }
+
+    fn apply<'t>(
+        &self,
+        descendants: Vec<MergedDescendant<'t>>,
+        deletions: Vec<Deletion>,
+    ) -> Result<()> {
+        if !self.has_changes()? {
+            return Ok(());
+        }
+        let tx = self.db.unchecked_transaction()?;
+        let result = self
+            .update_local_items(descendants, deletions)
+            .and_then(|_| self.stage_local_items_to_upload())
+            .and_then(|_| {
+                self.db.execute_batch(
+                    "
+                    DELETE FROM mergedTree;
+                    DELETE FROM idsToWeaklyUpload;",
+                )?;
+                Ok(())
+            });
+        match result {
+            Ok(_) => tx.commit()?,
+            Err(_) => tx.rollback()?,
```

Comment:
Not needed; `tx` will roll back automatically if it's not committed when dropped.